metasynthesis/abstract_genetic.py

```diff
@@ -0,0 +1,67 @@
+from typing import List, Callable, Tuple
+
+Genome = List[int]
+Population = List[Genome]
+PopulateFunc = Callable[[], Population]
+FitnessFunc = Callable[[Genome], float]
+SelectionFunc = Callable[[Population, FitnessFunc], Tuple[Genome, Genome]]
+CrossoverFunc = Callable[[Genome, Genome], Tuple[Genome, Genome]]
+MutationFunc = Callable[[Genome], Genome]
+
+
+class GeneticAlgorithm:
+    """Abstract interface for a genetic algorithm to synthesise elements used in program synthesis"""
+
+    def __init__(self, fitness_limit: int, generation_limit: int):
+        self.fitness_limit = fitness_limit
+        self.generation_limit = generation_limit
+
+    def generate_genome(self, length: int) -> Genome:
+        """This method creates a new genome of the specified length"""
+
+        raise NotImplementedError()
+
+    def generate_population(self, size: int, genome_length: int) -> Population:
```

Comment:
I believe that genome_length can be removed in order to allow for variable length encoding.

Reply:
It's a choice I guess, but I think leaving it out is indeed more flexible